tests/core/contracts/test_contract_constructor.py

```diff
@@ -0,0 +1,134 @@
+import pytest
+
+from eth_utils import (
+    decode_hex,
+)
+
+
+def test_contract_constructor_abi_encoding_with_no_constructor_fn(MathContract, MATH_CODE):
+    deploy_data = MathContract.constructor()._encode_data_in_transaction()
+    assert deploy_data == MATH_CODE
+
+
+def test_contract_constructor_gas_estimate_no_constructor(MathContract):
+    gas_estimate = MathContract.constructor().estimateGas()
+    try:
+        assert abs(gas_estimate - 21472) < 200  # Geth
+    except AssertionError:
+        assert abs(gas_estimate - 167412) < 200  # eth-tester with py-evm
+
+
+def test_contract_constructor_gas_estimate_with_constructor_without_arguments(
+        SimpleConstructorContract):
+    gas_estimate = SimpleConstructorContract.constructor().estimateGas()
+    try:
+        assert abs(gas_estimate - 21472) < 200  # Geth
+    except AssertionError:
+        assert abs(gas_estimate - 43082) < 200  # eth-tester with py-evm
+
+
+def test_contract_constructor_gas_estimate_with_constructor_with_arguments(
+        WithConstructorArgumentsContract):
+    gas_estimate = WithConstructorArgumentsContract.constructor(args=[1234, 'abcd']).estimateGas()
```

Comment:
It doesn't work current implementation since `constructor()`'s signature is
```
    def constructor(cls, args=None, kwargs=None)
```
And we will then get `args=1234, kwargs=b'abcd'`...